modules/core/src/directx.cpp

```diff
@@ -248,25 +248,106 @@ Context& initializeContextFromD3D11Device(ID3D11Device* pD3D11Device)
 #elif !defined(HAVE_OPENCL)
     NO_OPENCL_SUPPORT_ERROR;
 #else
-    cl_uint numPlatforms;
-    cl_int status = clGetPlatformIDs(0, NULL, &numPlatforms);
-    if (status != CL_SUCCESS)
-        CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get number of platforms");
-    if (numPlatforms == 0)
-        CV_Error(cv::Error::OpenCLInitError, "OpenCL: No available platforms");
-
-    std::vector<cl_platform_id> platforms(numPlatforms);
-    status = clGetPlatformIDs(numPlatforms, &platforms[0], NULL);
-    if (status != CL_SUCCESS)
-        CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get number of platforms");
-
-    // TODO Filter platforms by name from OPENCV_OPENCL_DEVICE
-
-    int found = -1;
-    cl_device_id device = NULL;
-    cl_uint numDevices = 0;
-    cl_context context = NULL;
-
+	cl_uint numPlatforms;
+	cl_int status = clGetPlatformIDs(0, NULL, &numPlatforms);
+	if (status != CL_SUCCESS)
+		CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get number of platforms");
+	if (numPlatforms == 0)
+		CV_Error(cv::Error::OpenCLInitError, "OpenCL: No available platforms");
+
+	std::vector<cl_platform_id> platforms(numPlatforms);
+	status = clGetPlatformIDs(numPlatforms, &platforms[0], NULL);
+	if (status != CL_SUCCESS)
+		CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get number of platforms");
+
+	// TODO Filter platforms by name from OPENCV_OPENCL_DEVICE.
+	int found = -1;
+	cl_device_id device = NULL;
+	cl_uint numDevices = 0;
+	cl_context context = NULL;
+
+#ifdef HAVE_CUDA
```

Comment:
Do we really need CUDA SDK for this?
I believe it is "OpenCL" feature and CUDA is not used here.

Reply:
The current situation is that the NVIDIA CUDA platform does not support the `cv_khr_d3d11_sharing` extension, so in order to support D3D11 interoperability on NVIDIA CUDA platform, we need to support the `cv_nv_d3d11_sharing` extension. Current changes ensure that users whose compilation options are selected for WITH_CUDA can use the NVIDIA CUDA platform, while unchecked ones do not affect the original functionality. Perhaps it would be a better way to determine whether an OCL device has a `cv_nv_d3d11_sharing` extension? But at the moment I'm not sure how to judge this at compile time. It would be great if you could give me some advice.

Reply:
What if I change it like this?
[`OpenCVDetectOpenCL.cmake`](https://github.com/opencv/opencv/blob/master/cmake/OpenCVDetectOpenCL.cmake):

```cpp
if(OPENCL_FOUND)
...
  try_compile(__VALID_CL_KHR_D3D11_SHARING
    "${OpenCV_BINARY_DIR}"
    "${OpenCV_SOURCE_DIR}/cmake/checks/opencl.cpp"
    COMPILE_DEFINITIONS "-DCHECK_CL_KHR_D3D11_SHARING"
    OUTPUT_VARIABLE TRY_OUT
  )
  if(__VALID_CL_KHR_D3D11_SHARING)
    set(HAVE_CL_KHR_D3D11_SHARING ON)
  endif()

  try_compile(__VALID_CL_NV_D3D11_SHARING
    "${OpenCV_BINARY_DIR}"
    "${OpenCV_SOURCE_DIR}/cmake/checks/opencl.cpp"
    COMPILE_DEFINITIONS "-DCHECK_CL_NV_D3D11_SHARING"
    OUTPUT_VARIABLE TRY_OUT
  )
  if(__VALID_CL_NV_D3D11_SHARING)
    set(HAVE_CL_NV_D3D11_SHARING ON)
  endif()
...
endif()
```

[`cmake/check/opencl.cpp`](https://github.com/opencv/opencv/blob/master/cmake/checks/opencl.cpp):

```cpp
#include <opencv2/core/ocl.hpp>
...
int main(int /*argc*/, char** /*argv*/)
{
#ifdef CHECK_CL_KHR_D3D11_SHARING
    bool khr_d3d11 = cv::ocl::Device::getDefault().isExtensionSupported("cl_khr_d3d11_sharing");
#endif

#ifdef CHECK_CL_NV_D3D11_SHARING
    bool nv_d3d11 = cv::ocl::Device::getDefault().isExtensionSupported("cl_nv_d3d11_sharing");
#endif
    return 0;
}
```

Reply:
The build log display buildbot also does not support the `cl_khr_d3d11_sharing` extension.

Reply:
`cl_khr_d3d11_sharing` is available with AMD devices (probably on Windows). Currently we don't have such build workers.